Beginnings of a CPP implementation

ValHayot opened the pull request.

arokem commented:
Thanks for putting in this WIP here for comments! One initial question: would it be hard to separate the Python stuff (testing, types, documentation) from the CPP stuff? It would be great to have the Python stuff on a separate PR (and probably relatively uncontroversial to merge into the existing prototype).

ValHayot commented:
Not at all! Sorry, I neglected to realize this. Will update PR.

frankyeh commented:
Great implementation!!

Here are my thoughts at the first glimpse about potential challenges:

* inclusion of <filesystem> needs compiler's full support of C++14. clang under the majority of MacOS has not supported it yet.
* the dependencies on the other 6 libraries, including PkgConfig, ZLIB, libzip, Eigen, jsoncpp, GTest can be a lot of work to cross plateform.  Some of them may not have readily available Windows libraries. I see the necessity of using zlib, libzip, jsoncpp, but not sure if PkgConfig, Eigen, Gtest is needed here for the core function.
* The buffer in Json::Value load_header is a naked pointer and not freed on all existing points. 
* zip_file_t *zh in Json::Value load_header is also a naked pointer and not freed.
* zip_t *zf in in int load_from_zip has the same problem.
* template <typename Derived> implementation in cpp file is going to be a problem for a lot of compilers. For best portability, any template implementation should be in the header file.
* the TrxFile class is used in two functions as the return type but does not implement a move constructor. 
* replace NULL by nullptr

baranaydogan commented:
Thank you for initiating the C++ implementation. Great job!

+1 for all the points mentioned by @frankyeh above.
+10 for his point about dependencies.

frheault commented:
@Lestropie @jdtournier is it possible to have your input on the libraries, the compatibility requirements or ''macro'' development choice? That would help to have a working c++ prototype that already looks close to what other tools have, and avoid back & forth  later on.

Feedback would be very useful

jdtournier commented:
Apologies I've not had any input so far - always struggling to find time. It'll be a while before I can go through the code properly, but just a couple of comments based on @frankyeh's earlier post:

- If filesystem requires c++14, that might be a problem for MRtrix - we still rely on C++11. However, given how C++14 really isn't a massive upgrade from C++11, this might not be an issue for long...
- In terms of dependencies, I would also recommend keeping them to the absolute minimum. `libzip` is clearly going to be essential. But some of the others might not be a problem:
  - zlib is available on all platforms - certainly pretty much all platforms I've ever had to deal with
  - Eigen is header-only, and hence trivial to integrate into any C++ project
  - pkconfig a compile-time helper tool. If it's not available, it should always be possible to provide the correct flags for each dependency manually
  - for JSON, the current jsoncpp dependencies seems to require a library, and that may be a problem. In MRtrix, we use [JSON for modern C++](https://github.com/nlohmann/json), which is very lightweight, and more importantly header-only, and hence trivial to integrate into any C++ project. 

I'll have a look through the code when I get the chance (not likely for the next two weeks given Easter holidays - sorry!).

ValHayot commented:
Hey everyone,

Thanks for all the feedback so far! I'm still in the process of implementing functions and addressing your recommendations. However, at the time being, I have: 
- Reverted the code to C++11 (I'm using C POSIX functions instead and am making assumptions on fs paths which would definitely not be compatible with Windows for the time being)
- Updated the JSON library to be JSON for modern C++
- Cleaned up some of the dependencies.
- Implemented additional functions and added tests for those

I'll hopefully have more complete code soon.

The python code has a function to build a TRX file from a reference (Trk, NifTI) which would require me to either implement the code or use a library. Any suggestions on what would be the best approach and/or suggested libraries ?

Lestropie commented on trx_file_memmap/cpp/src/trx.cpp:
```diff
@@ -0,0 +1,361 @@
+using namespace Eigen;
+using namespace std;
+
+std::string get_base(const std::string &delimeter, const std::string &str)
```
"delimiter"

anibalsolon commented on trx_file_memmap/cpp/src/trx.cpp:
```diff
@@ -0,0 +1,361 @@
+	{
+
+		// TODO: won't work on windows and not validating OS type
+		std::string base = get_base("/", filename);
```
Since we're using C++11, I wonder if setting a constant so we could use it in the project would be better, not requiring parameterizing the delimiter.

e.g.

```
#if defined(WIN32) || defined(_WIN32) 
#define PATH_SEPARATOR "\\" 
#else 
#define PATH_SEPARATOR "/" 
#endif 
```

ValHayot replied:
That's a good idea! I was trying to defer to built-in functions for handling path difference, but use of compiler directives would be useful either way and this is a good way to get things working on Windows for now.

anibalsolon commented on trx_file_memmap/cpp/src/trx.cpp:
```diff
@@ -0,0 +1,361 @@
+
+			file_pointer_size[elem_filename] = {mem_address, global_pos};
+		}
+		return 1;
```
Should it return 0 ?

ValHayot replied:
You're right, it's also not supposed to return an int. I'll update the code shortly!

anibalsolon commented on trx_file_memmap/cpp/src/trx.cpp:
```diff
@@ -0,0 +1,361 @@
+			file.seekg(global_pos);
+			mem_address = global_pos;
+
+			unsigned char signature[4] = {0};
+			const unsigned char local_sig[4] = {0x50, 0x4b, 0x03, 0x04};
+			file.read((char *)signature, sizeof(signature));
+
+			if (memcmp(signature, local_sig, sizeof(signature)) == 0)
+			{
+				global_pos += 30;
+				global_pos += sb.comp_size + elem_filename.size();
+			}
```
I may not have completely understood this part- so if the file within the .trx is a zip file, it should skip the zip header? wouldn't it be relevant information for libzip or wouldn't libzip/trxlib assume that the file cursor starts at 0?

anibalsolon commented on trx_file_memmap/cpp/src/trx.tpp:
```diff
@@ -0,0 +1,685 @@
+	zip_t *zf = zip_open(filename.c_str(), 0, errorp);
+	json header = load_header(zf);
+
+	std::map<std::string, std::tuple<int, int>> file_pointer_size;
```
The file size/offset may need to be greater than `int`, I am having a problem here loading files >2GB.  Afaik they are positive, so `uint` gives a bit more of a range, but `long long` may cover other larger cases. Changing it (and downstream in fn calls) to `long long` worked here.

ValHayot replied:
Sounds good! Would you like to add your updates?

arokem commented:
This PR should be merged now (?), and then the code will be moved to the https://github.com/tee-ar-ex/trx-cpp repository for further development.

arokem commented:
I'm going to fix the conflict that was created with the merge of #27
